log.py

```diff
@@ -0,0 +1,161 @@
+import aiofiles
+import asyncio
+import json
+import os
+
+from aiofiles.threadpool.text import AsyncTextIOWrapper
+from enum import Enum
+from dataclasses import dataclass, field
+from typing import Self
+
+from rpc.raft_pb2 import AppendEntriesRequest, LogEntry
+from task_scheduler import run_in_background
+
+
+class JSONCodec:
+    class Field(Enum):
+        KEY = "key"
+        VALUE = "value"
+        TERM = "term"
+        INDEX = "index"
+
+    @staticmethod
+    def decode_log_entry(d: dict[str, str]) -> LogEntry:
+        key: str = d[JSONCodec.Field.KEY.value]
+        value: str = d[JSONCodec.Field.VALUE.value]
+        term: int = int(d[JSONCodec.Field.TERM.value])
+        index: int = int(d[JSONCodec.Field.INDEX.value])
+        return LogEntry(term=term, index=index, key=key, value=value)
+
+    @staticmethod
+    def encode_log_entry(entry: LogEntry) -> str:
+        return json.dumps(
+            {
+                JSONCodec.Field.INDEX.value: entry.index,
+                JSONCodec.Field.TERM.value: entry.term,
+                JSONCodec.Field.KEY.value: entry.key,
+                JSONCodec.Field.VALUE.value: entry.value,
+            }
+        )
+
+
+@dataclass(slots=True)
+class Log:
+    _path: str = field()
+    _commit_index: int = field(default=-1)
+    _fp: AsyncTextIOWrapper = field(init=False)
+    _last_applied: int = field(init=False)
+    _entries: list[LogEntry] = field(default_factory=list)
+    _cur_state: dict[str, str] = field(default_factory=dict)
+    _apply_entries_task: asyncio.Task | None = field(default=None)
+
+    @classmethod
+    async def create(cls, path: str) -> Self:
+        ret: Self = cls(path)
```

Comment:
is it possible to build this up immutably instead of creating it and assigning to it?